initgeneration/initgeneration.go

```diff
@@ -0,0 +1,247 @@
+	if err != nil {
+		return nil, err
+	}
+
+	script, err := g.prepareScriptForDynaKube(dk, kubeSystemUID, infraMonitoringNodes)
+	if err != nil {
+		return nil, err
+	}
+
+	data, err := script.generate()
+	if err != nil {
+		return nil, err
+	}
+
+	return data, nil
+}
+
+func (g *InitGenerator) prepareScriptForDynaKube(dk *dynatracev1alpha1.DynaKube, kubeSystemUID types.UID, infraMonitoringNodes map[string]string) (*script, error) {
+	var tokens corev1.Secret
+	if err := g.client.Get(context.TODO(), client.ObjectKey{Name: dk.Tokens(), Namespace: g.namespace}, &tokens); err != nil {
+		return nil, errors.WithMessage(err, "failed to query tokens")
+	}
+
+	var proxy string
+	if dk.Spec.Proxy != nil {
+		if dk.Spec.Proxy.ValueFrom != "" {
+			var ps corev1.Secret
+			if err := g.client.Get(context.TODO(), client.ObjectKey{Name: dk.Spec.Proxy.ValueFrom, Namespace: g.namespace}, &ps); err != nil {
+				return nil, fmt.Errorf("failed to query proxy: %w", err)
+			}
+			proxy = string(ps.Data[proxyInitSecretField])
+		} else if dk.Spec.Proxy.Value != "" {
+			proxy = dk.Spec.Proxy.Value
+		}
+	}
+
+	var trustedCAs []byte
+	if dk.Spec.TrustedCAs != "" {
+		var cam corev1.ConfigMap
+		if err := g.client.Get(context.TODO(), client.ObjectKey{Name: dk.Spec.TrustedCAs, Namespace: g.namespace}, &cam); err != nil {
+			return nil, fmt.Errorf("failed to query ca: %w", err)
+		}
+		trustedCAs = []byte(cam.Data[trustedCASecretField])
+	}
+
+	return &script{
+		ApiUrl:        dk.Spec.APIURL,
+		SkipCertCheck: dk.Spec.SkipCertCheck,
+		PaaSToken:     string(tokens.Data[dtclient.DynatracePaasToken]),
+		Proxy:         proxy,
+		TrustedCAs:    trustedCAs,
+		ClusterID:     string(kubeSystemUID),
+		TenantUUID:    dk.Status.ConnectionInfo.TenantUUID,
+		IMNodes:       infraMonitoringNodes,
+	}, nil
+}
+
+// getInfraMonitoringNodes creates a mapping between all the nodes and the tenantUID for the infra-monitoring dynakube on that node.
+// Possible mappings:
+// - mapped: there is a infra-monitoring agent on the node, and the dynakube has the tenantUID set => user processes will be grouped to the hosts  (["node.Name"] = "dynakube.tenantUID")
+// - not-mapped: there is NO infra-monitoring agent on the node => user processes will show up as individual 'fake' hosts (["node.Name"] = "-")
+// - unknown: there SHOULD be a infra-monitoring agent on the node, but dynakube has NO tenantUID set => user processes will restart until this is fixed (node.Name not present in the map)
+//
+// Checks all the dynakubes with infra-monitoring against all the nodes (using the nodeSelector), creating the above mentioned mapping.
+func (g *InitGenerator) getInfraMonitoringNodes(dk *dynatracev1alpha1.DynaKube) (map[string]string, error) {
+	var dks dynatracev1alpha1.DynaKubeList
+	if err := g.client.List(context.TODO(), &dks, client.InNamespace(g.namespace)); err != nil {
+		return nil, errors.WithMessage(err, "failed to query DynaKubeList")
+	}
+
+	nodes, imNodes, err := g.initIMNodes()
```

Comment:
consider returning a struct containing nodes and imNodes instead of having three return values